### PR TITLE
Fix version number before and after 8.0

### LINKED ISSUE #26756: BO - Translations page - an exception is displayed
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

In the develop branch, in the BO > International > Translations page, an exception is displayed
image
It is detected by nightly
https://nightly.prestashop.com/report/3037
Version 1.8.0.0 is not correct.

### Expected behavior

_No response_

### Steps to reproduce

1. Go to BO > International > Translations page 
2. See error

### PrestaShop version(s) where the bug happened

develop

### PHP version(s) where the bug happened

7.3

### If your bug is related to a module, specify its name and its version

_No response_

### LINKED ISSUE #26760: BO - the version of PrestaShop in the toolbar symfony is not correct
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

In the BO > Migrated page, when the debug mode is enabled.
The version of PrestaShop in the toolbar is incorrect.
image
This issue is reported by @boubkerbribri 

### Expected behavior

It should be 8.0.0

### Steps to reproduce

1. Enable debug mode
2. Go to BO > Any migrated page
3. In the toolbar Symfony
4. check the version PrestaShop

### PrestaShop version(s) where the bug happened

develop

### PHP version(s) where the bug happened

7.3

### If your bug is related to a module, specify its name and its version

_No response_

### PR DESCRIPTION
**Pull request by sowbiba**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Version number returned depends on whether the pattern is 1.X or not. If the version given is 1.7.7.2 for example, the full version is 1.7.7.2, the major string is 1.7, the major number is 7, minor is 7 and patch is 2.<br>If the version is 8.1.2, majorString is 8, major is 8, minor 1 and patch 2.<br><br>This will fix the translation languages load as described in #26756 and the toolbar version bug in #26760 
| Type?             | bug fix
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #26756 and #26760 
| How to test?      | Follow descriptions in the issues linked
| Possible impacts? | -

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/26789)

### CODE REVIEW

File: tests/Unit/Core/Foundation/VersionTest.php [hunk 1]
```diff
@@ -442,10 +442,10 @@ public function provideVersions()
             '7.6.0' => [
                 '7.6.0',
                 [
-                    'version' => '1.7.6.0',
-                    'fullVersion' => '1.7.6.0',
+                    'version' => '7.6.0',
```
Comment (matks) [inline tests/Unit/Core/Foundation/VersionTest.php]: Interesting.

We could say this test is _wrong_ because you changed the expected result.

But 7.6.0 indeed matches SemVer and not legacy style. Previous implementation would match it to legacy style although it made no sense.

So we have an issue
- do we change this behavior because it's better this way?
- do we keep the behavior of before because we're afraid of changing this that might be used by modules (so we keep wrong behavior of the past to provide backward compatibility)

Comment (jolelievre) [reply tests/Unit/Core/Foundation/VersionTest.php]: I agree with the idea that `7.6.*` are correct semver and should have been handled correctly historically Though I have the same doubts regarding changing this behaviour I'm not sure about the impact this all could have

Previously major versions were not indicated in the first digit, that's a weird way of handling versions but maybe we should keep this in our history of code as well. So my suggestion would be not to change this behaviour, and the Version class should handle the fact that for versions prior to the 8 (so up until the 7) the version must be written with a prefix `1.`

Which means that in the PrestaShop version management `1.7.5` is equal to `7.5`, equal in the sense that those two inputs will identify the same version which will actually be transformed back or converted to `1.7.5` in all cases.

Comment (jolelievre) [reply tests/Unit/Core/Foundation/VersionTest.php]: So I think it would mean adding a condition in the Version class:

```php
$isLegacy = $major < 8;
```

File: src/Core/Foundation/Version.php [hunk 2]
```diff
@@ -391,4 +393,14 @@ private function removeLegacyPrefix($version, $majorVersionString)
 
         return $version;
     }
+
+    /**
+     * Checks if the version number matches old PS version styles (1.X.X)
+     */
+    private static function matchLegacyPattern(string $version): bool
+    {
+        $result = preg_match('/^1\.(.*)$/', $version); // version starting with "1."
```
Comment (PierreRambaud) [inline src/Core/Foundation/Version.php]: Seems to be false, `1.2` or `1.2.3` are not legacy pattern, the patter is four digits beginning with one :thinking:

File: tests/Unit/Core/Foundation/VersionTest.php (hunk 1 above)
Comment (PierreRambaud) [reply tests/Unit/Core/Foundation/VersionTest.php]: Maybe the legacy pattern check is not the right one :thinking:

File: src/Core/Foundation/Version.php (hunk 2 above)
Comment (sowbiba) [reply src/Core/Foundation/Version.php]: it's an understanding issue here

For me the old way to define versions (legacy) was wrong because we didn't increment the good digit on every major

after 1.0, the next major would be 2.0 and not 1.1

So I assume that anyversion number starting with `1.` is the legacy way of defining version

File: tests/Unit/Core/Foundation/VersionTest.php (hunk 1 above)
Comment (sowbiba) [reply tests/Unit/Core/Foundation/VersionTest.php]: Maybe ...


We can mix the 2 ways to determinate if it's a legacy version :

1 - If I have a 4 digits version 1.7.6.0 for example. Major = 7, minor = 6, patch = 0
2 - If I have a major version lower than 8.<br>7.6.0 will be equal to 1.7.6.0. Major = 7, minor = 6, patch = 0

Comment (jolelievre) [reply tests/Unit/Core/Foundation/VersionTest.php]: Maybe we could uniformize this:
- split the input version into 4 parts (when available use 0 as fallback for missing parts)
- if first part is `1` it is the legacy writing so we need to offset the index for each part
  - major is `$matches[2]`
  - minor is `$matches[3]`
  - patch is `$matches[4]`
- for other cases it means the input version respects the usual semver notation so we don't need to ignore the first match:
  - major is `$matches[1]`
  - minor is `$matches[2]`
  - patch is `$matches[3]`
- as for the output (aka fullVersion aka $this->version) as long as the major is < 8 we always display with the legacy style `1.{major}.{minor}.{patch}` (it is not correct semver but that was our legacy way 🤷‍♂️ )

Comment (jolelievre) [reply tests/Unit/Core/Foundation/VersionTest.php]: An alternative could also be to have two different getters, one that handles this switch between legacy/semver and another one which always respect semver (so it would return `7.5` for a `1.7.5` input)

Comment (jolelievre) [reply tests/Unit/Core/Foundation/VersionTest.php]: (Oh actually we already have this with `getSemVersion()` my bad)

File: src/Core/Foundation/Version.php (hunk 2 above)
Comment (matks) [reply src/Core/Foundation/Version.php]: I agree with @PierreRambaud : we should consider legacy version numbers need 4 digits

File: tests/Unit/Core/Foundation/VersionTest.php (hunk 1 above)
Comment (matks) [reply tests/Unit/Core/Foundation/VersionTest.php]: The suggestion of @jolelievre seems very good to me 👍

File: src/Core/Foundation/Version.php (hunk 2 above)
Comment (jolelievre) [reply src/Core/Foundation/Version.php]: I'm not sure the four numbers should be mandatory, for example `1.7.5` is a legacy version even if the patch part is missing

Comment (PierreRambaud) [reply src/Core/Foundation/Version.php]: When you say it's a legacy verison number, it's the four digits version (x.x.x.x), otherwise it's not a legacy version, it's just following semver.
Make this Version class working on modules/Core, it costs nothing to check the four digits and be sure it targets the old PrestaShop way.

Comment (eternoendless) [reply src/Core/Foundation/Version.php]: Pierre is right

File: src/Core/Foundation/Version.php [hunk 3]
```diff
@@ -162,14 +168,12 @@ public static function buildFromString($version)
...
-        $version = '1.' . $this->version;
-
         if (!$full) {
             // remove extra parts
-            return preg_replace('/[-\+].*/', '', $version);
+            return preg_replace('/[-\+].*/', '', $this->version);
         }
 
-        return $version;
+        return $this->version;
```
Comment (eternoendless) [inline src/Core/Foundation/Version.php:186]: You're making `version` (the legacy result) the same as `semVersion`, which is a breaking change and might produce unforeseen consequences. This change is also not aligned with the method's comment.

File: tests/Unit/Core/Foundation/VersionTest.php [hunk 4]
```diff
@@ -512,17 +512,31 @@ public function provideVersions()
             '8.1.2' => [
                 '8.1.2',
                 [
-                    'version' => '1.8.1.2',
-                    'fullVersion' => '1.8.1.2',
+                    'version' => '8.1.2',
```
Comment (PierreRambaud) [inline tests/Unit/Core/Foundation/VersionTest.php:529]: we should keep legacy behavior too, wdyt? :thinking:

Comment (sowbiba) [reply tests/Unit/Core/Foundation/VersionTest.php:529]: this is where the bug is :D

8+ versions should not be converted to 1.8+

Comment (PierreRambaud) [reply tests/Unit/Core/Foundation/VersionTest.php:529]: We should keep the behavior for `1.8.0.1` and `8.0.1` which are different versions

Comment (sowbiba) [reply tests/Unit/Core/Foundation/VersionTest.php:529]: I added a test for 1.8+

**Review by matks — APPROVED** [review]

**Comment by matks** [comment]
@sowbiba please check changes in https://github.com/PrestaShop/PrestaShop/pull/26867

File: tests/Unit/Core/Foundation/VersionTest.php [hunk 5]
```diff
@@ -509,8 +509,8 @@ public function provideVersions()
                     'buildMeta' => 'nightly.20190526',
                 ],
             ],
-            '8.1.2' => [
-                '8.1.2',
+            '1.8.1.2' => [
+                '1.8.1.2',
```
Comment (PierreRambaud) [inline tests/Unit/Core/Foundation/VersionTest.php:513]:
```suggestion
            '8.1.2' => [
                '8.1.2',
```

It should be that because the version passed is 1.8.1.2 like above with `1.7.7.0`

File: src/Core/Foundation/Version.php [hunk 6]
```diff
@@ -123,19 +123,35 @@ public function __construct(
...
 
+        $isLegacy = false;
         if (isset($matches['legacy']) && '' !== $matches['legacy']) {
             // legacy version like "1.7.5.0"
+            $isLegacy = true;
             $major = (int) $matches[2];
             $minor = (int) $matches[3];
             $patch = (int) $matches['legacy'];
-            $version = substr($version, strpos($version, '.') + 1);
+        } elseif (8 > (int) $matches[1]) {
```
Comment (PierreRambaud) [inline src/Core/Foundation/Version.php:149]: Why?
What's the relationship with 8?

File: src/Core/Foundation/Version.php [hunk 7]
```diff
@@ -123,19 +123,35 @@ public function __construct(
...
+            $isLegacy = true;
+            $major = (int) $matches[1];
+            $minor = isset($matches[2]) ? (int) $matches[2] : '0';
+            $patch = isset($matches[3]) ? (int) $matches[3] : '0';
+            $version = "1.$version";
```
Comment (PierreRambaud) [inline src/Core/Foundation/Version.php:154]: It's not a legacy version, so why did you add `1`?

File: src/Core/Foundation/Version.php (hunk 3 above)
Comment (PierreRambaud) [reply src/Core/Foundation/Version.php:186]: :+1:

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
Few comments ;)

**Comment by eternoendless** [comment]
Guys, this class works as expected. You're trying to fix the issue in the wrong place.

This class takes a version string, and works like this:
- If there are four components, it's the legacy format, so the first pair is thrown away
- If there are three or less components, it's a standard format

Whatever happens, getVersion() adds 1 before the major version. It's meant to return the legacy format.
To get the standard format, you use `getSemVersion()`. That's all.

**Comment by sowbiba** [comment]
Replaced by https://github.com/PrestaShop/PrestaShop/pull/26934
